src/parquet/file/metadata.h

```diff
@@ -0,0 +1,190 @@
+  int64_t distinct_count;
+  const std::string* min;
+  const std::string* max;
+};
+
+class PARQUET_EXPORT ColumnMetaData {
+ public:
+  // API convenience to get a MetaData accessor
+  static std::unique_ptr<ColumnMetaData> GetColumnMetaData(const uint8_t* metadata);
+
+  explicit ColumnMetaData(const uint8_t* metadata);
+  ~ColumnMetaData();
+
+  // column chunk
+  int64_t file_offset() const;
+  const std::string& file_path() const;
+  // column metadata
+  Type::type type() const;
+  int64_t num_values() const;
+  std::shared_ptr<schema::ColumnPath> path_in_schema() const;
+  bool is_stats_set() const;
+  ColumnStatistics Stats() const;
+  Compression::type compression() const;
+  std::vector<Encoding::type> Encodings() const;
+  int64_t total_compressed_size() const;
+  int64_t total_uncompressed_size() const;
+
+ private:
+  // PIMPL Idiom
+  class ColumnMetaDataImpl;
+  std::unique_ptr<ColumnMetaDataImpl> impl_;
+};
+
+class PARQUET_EXPORT RowGroupMetaData {
+ public:
+  // API convenience to get a MetaData accessor
+  static std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(const uint8_t* metadata);
+
+  explicit RowGroupMetaData(const uint8_t* metadata);
+  ~RowGroupMetaData();
+
+  // row-group metadata
+  int num_columns() const;
+  int64_t num_rows() const;
+  int64_t total_byte_size() const;
+  std::unique_ptr<ColumnMetaData> GetColumnMetaData(int i) const;
+
+ private:
+  // PIMPL Idiom
+  class RowGroupMetaDataImpl;
+  std::unique_ptr<RowGroupMetaDataImpl> impl_;
+};
+
+class PARQUET_EXPORT FileMetaData {
+ public:
+  // API convenience to get a MetaData accessor
+  static std::unique_ptr<FileMetaData> GetFileMetaData(const uint8_t* metadata);
+
+  explicit FileMetaData(const uint8_t* metadata);
+  ~FileMetaData();
+
+  // file metadata
+  int64_t num_rows() const;
+  int num_row_groups() const;
+  int32_t version() const;
+  const std::string& created_by() const;
+  int num_schema_elements() const;
+  std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(int i) const;
+
+  void WriteTo(OutputStream* dst);
+
+  // Return const-pointer to make it clear that this object is not to be copied
+  const SchemaDescriptor* schema() const;
+
+ private:
+  // PIMPL Idiom
+  class FileMetaDataImpl;
+  std::unique_ptr<FileMetaDataImpl> impl_;
+};
+
+// Builder API
+class PARQUET_EXPORT ColumnMetaDataBuilder {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<ColumnMetaDataBuilder> GetColumnMetaDataBuilder(
+      const std::shared_ptr<WriterProperties>& props, const ColumnDescriptor* column,
+      uint8_t* contents);
+
+  explicit ColumnMetaDataBuilder(const std::shared_ptr<WriterProperties>& props,
+      const ColumnDescriptor* column, uint8_t* contents);
+  ~ColumnMetaDataBuilder();
+
+  // column chunk
+  // Used when a dataset is spread across multiple files
+  void set_file_path(const std::string& path);
+  // column metadata
+  void SetStatistics(const ColumnStatistics& stats);
```

Comment:
As mentioned before: A comment about the ownership semantics of min/max would be nice.